[Synthetics UI] Add or edit a monitor

dominiqueclarke opened the pull request:
## Summary

Resolves https://github.com/elastic/kibana/issues/128473
Relates to https://github.com/elastic/kibana/issues/120638

<img width="1454" alt="Screen Shot 2022-06-27 at 11 11 50 AM" src="https://user-images.githubusercontent.com/11356435/175974074-d38ac52d-d1bb-48d9-b4a8-77a28fe167a3.png">
<img width="1455" alt="Screen Shot 2022-06-27 at 10 57 43 AM" src="https://user-images.githubusercontent.com/11356435/175973984-06b987e8-50fe-416e-ac08-b67dafa0446e.png">


Adds the monitor add and edit page.

@colleenmcginnis to confirm content

### Testing
--
This issue is very detailed. The best way to test is to review the AC listed in https://github.com/elastic/kibana/issues/128473 against the implementation

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/advanced/index.tsx:
```diff
@@ -0,0 +1,40 @@
+import { Field } from '../form/field';
+import { FormMonitorType } from '../types';
+
+export const AdvancedConfig = () => {
```
I will be editing this to take props.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/code_editor.tsx:
```diff
@@ -0,0 +1,58 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/combo_box.tsx:
```diff
@@ -0,0 +1,79 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/header_field.tsx:
```diff
@@ -0,0 +1,77 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/index_response_body_field.tsx:
```diff
@@ -0,0 +1,101 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/key_value_field.tsx:
```diff
@@ -0,0 +1,201 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/request_body_field.tsx:
```diff
@@ -0,0 +1,208 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/script_recorder_fields.tsx:
```diff
@@ -0,0 +1,133 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/source_field.tsx:
```diff
@@ -0,0 +1,159 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/uploader.tsx:
```diff
@@ -0,0 +1,94 @@
+/*
```
Pulled over from legacy uptime, with a few adjustments to make it work with react-hook-form.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+  ),
+};
+
+export const ADD_MONITOR_STEPS: StepMap = {
```
Configuration for monitor steps passed to `EuiSteps`.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+  },
+};
+
+export const FIELD: Record<string, FieldMeta> = {
```
Configuration for each individual field, mapping them back to each key and defining the rendered component as well as additional props, validation, error messages, etc.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+    required: true,
+    component: MonitorTypeRadioGroup,
+    ariaLabel: 'Monitor Type',
+    controlled: true,
```
Whether or not the component requires a react-hook-form `Controller`

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+        {
+          "Change the default namespace. This setting changes the name of the monitor's data stream. "
+        }
+        <EuiLink href="#">Learn More</EuiLink>
```
Need to get the link.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/context.tsx:
```diff
@@ -0,0 +1,25 @@
+/*
```
This is not supposed to be here. Deleting.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/render_app.tsx:
```diff
@@ -41,6 +41,8 @@ export function renderApp(
 
   const canSave = (capabilities.uptime.save ?? false) as boolean; // TODO: Determine for synthetics
 
+  console.warn('render stuff');
```
```suggestion
```

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+    fieldKey: ConfigKey.HOSTS,
+    required: true,
+    component: EuiFieldText,
+    label: 'Host:Port',
```
This should include required validation.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+    fieldKey: ConfigKey.HOSTS,
+    required: true,
+    component: EuiFieldText,
+    label: 'Host',
```
This should include required validation

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+    required: true,
+    component: EuiFieldText,
+    label: 'URL',
+    helpText: 'For example, your service endpoint.',
```
This should include required validation. I guess if I have the top level required prop (which is used to determine whether or not to show the optional tag), I can include the required validation automatically.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,819 @@
+      required: true,
+    }),
+  },
+  [`${ConfigKey.URLS}__single`]: {
```
This needs required validation.

dominiqueclarke commented on x-pack/plugins/synthetics/e2e/page_objects/synthetics_app.tsx:
```diff
@@ -63,9 +62,31 @@ export function syntheticsAppPageProvider({ page, kibanaUrl }: { page: Page; kib
+        await page.click('text=delete', { delay: 800 });
+        await page.waitForSelector('[data-test-subj="confirmModalTitleText"]');
+        await this.clickByTestSubj('confirmModalConfirmButton');
+        // isSuccessful = Boolean(await this.findByTestSubj('uptimeDeleteMonitorSuccess'));
```
I was having trouble with monitor deletion, but I think it's been resolved. Need to update this.

elasticmachine commented:
Pinging @elastic/uptime (Team:uptime)

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/uploader.tsx:
```diff
@@ -0,0 +1,99 @@
+);
+
+const PROMPT_TEXT = i18n.translate(
+  'xpack.synthetics.createPackagePolicy.stepConfigure.monitorIntegrationSettingsSection.browser.uploader.label',
```
```suggestion
  'xpack.synthetics.monitorConfig.uploader.label',
```

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,1420 @@
+        <>
+          <p>
+            <FormattedMessage
+              id="xpack.synthetics.monitorConfig.monitorScriptEditStep.descriptio"
```
```suggestion
              id="xpack.synthetics.monitorConfig.monitorScriptEditStep.description"
```

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,1420 @@
+          <p>
+            <FormattedMessage
+              id="xpack.synthetics.monitorConfig.monitorScriptEditStep.descriptio"
+              defaultMessage="Use Elastic Script Recorder to generate and upload a script. Alternatively, you can edit the existing Playwright script (or paste a new one) in the script editor."
```
```suggestion
              defaultMessage="Use Elastic Script Recorder to generate and upload a script. Alternatively, you can edit the existing {playwright} script (or paste a new one) in the script editor."
```

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/routes.tsx:
```diff
@@ -159,26 +172,67 @@ const getRoutes = (euiTheme: EuiThemeComputed): RouteProps[] => {
     },
     {
       title: i18n.translate('xpack.synthetics.addMonitorRoute.title', {
```
```suggestion
      title: i18n.translate('xpack.synthetics.createMonitorRoute.title', {
```

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/routes.tsx:
```diff
@@ -158,27 +171,68 @@ const getRoutes = (euiTheme: EuiThemeComputed): RouteProps[] => {
       pageHeader: {
         pageTitle: (
           <FormattedMessage
             id="xpack.synthetics.addMonitor.pageHeader.title"
```
```suggestion
            id="xpack.synthetics.createMonitor.pageHeader.title"
```

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/controlled_field.tsx:
```diff
@@ -0,0 +1,91 @@
+      isInvalid={isInvalid}
+      error={isInvalid ? hookError || fieldState.error?.message || error : undefined}
+    >
+      <Component
```
Disambiguate `Component` from React's exported `Component` with something local to context e.g. `FieldComponent` or better.

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/controlled_field.tsx:
```diff
@@ -0,0 +1,91 @@
+  component: Component,
+  props,
+  fieldKey,
+  useSetValue,
```
Would it be better to disambiguate `useSetValue` to not be taken as a hook, i.e. to booleanize the prop name to `shouldUseSetValue`, `isUseSetValue` (or better)?

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/use_monitor_name.ts:
```diff
@@ -0,0 +1,64 @@
+/*
```
The file can be moved up one level into the **hooks** directory.

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,1420 @@
+  },
+};
+
+export const FIELD_CONFIG: FieldConfig = {
```
Only group of related constants and defs to a separate file. It seems stuff on L#81 - L#214 can have it's own file. Besides that it looks good now. Also see if we can do without exporting the `HTTP_ADVANCED` and `TCP_ADVANCED` if not needed outside of config.tsx.

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,1385 @@
+      'xpack.synthetics.monitorConfig.monitorType.multiStep.description',
+      {
+        defaultMessage:
+          'A lightweight API check to validate the availability of a web service or endpoint.',
```
Change copy.

dominiqueclarke replied:
Adjusted

awahab07 commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/config.tsx:
```diff
@@ -0,0 +1,1385 @@
+      'xpack.synthetics.monitorConfig.monitorType.singlePage.description',
+      {
+        defaultMessage:
+          'A lightweight API check to validate the availability of a web service or endpoint.',
```
Change copy. Also see if copies for TCP and ICMP are correct.

dominiqueclarke replied:
Adjusted. @colleenmcginnis Do you have specific copy for the descriptions for tcp and icmp monitors.

awahab07 commented:
There's a small glitch with the validation state of Locations field. 

Click **Create Monitor** with empty Locations. Locations field gets into error state but stay red despite providing a valid value.

![Screenshot 2022-06-30 at 13 47 18](https://user-images.githubusercontent.com/2748376/176671118-1cd01509-cb11-4b25-963c-823c85b982fa.png)

dominiqueclarke commented:
@elasticmachine merge upstream

awahab07 commented:
@elasticmachine merge upstream

awahab07 reviewed (approved):
It LGTM. Some of the suggested refactors could be made in follow-ups.

There are a couple functionality gotchas which could also be adjusted as follow-ups:
1. Monitor Name is bound to Website URL (or URL, Host, Host:Port) which is fine if it's intentional. In case of Single Page monitor, the Website URL gets empty when monitor is edited the second time.

https://user-images.githubusercontent.com/2748376/178241417-b3d5b621-c4c2-408b-8577-c4f5c6d98f53.mov

2. Updating a disabled monitor enables it.

https://user-images.githubusercontent.com/2748376/178241639-3556ae73-937a-4ff6-b7a8-fa57b5040a1a.mov

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/types.ts:
```diff
@@ -0,0 +1,61 @@
+
+export type StepMap = Record<FormMonitorType, Step[]>;
+
+export * from '../../../../../common/runtime_types/monitor_management';
+export * from '../../../../../common/types/monitor_validation';
```
What is the concern here? I wanted to avoid super long import paths.

shahzad31 replied:
i think it could be an IDE thing. In webstorm, this leads to add odd DX, since clicking on type will lead me to this file instead of where it's actually declared.

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,1021 @@
+  [ConfigKey.PROXY_URL]: {
+    fieldKey: ConfigKey.PROXY_URL,
+    component: EuiFieldText,
+    label: 'Proxy URL',
```
i18n

dominiqueclarke commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/form_config.tsx:
```diff
@@ -0,0 +1,232 @@
+      },
+    ],
+  },
+  [FormMonitorType.SINGLE]: {
+    step1: [FIELD[ConfigKey.FORM_MONITOR_TYPE]],
+    step2: [
+      FIELD[`${ConfigKey.URLS}__single`],
+      FIELD[ConfigKey.TEXT_ASSERTION],
+      FIELD[ConfigKey.NAME],
+      FIELD[ConfigKey.LOCATIONS],
+      FIELD[ConfigKey.SCHEDULE],
+      FIELD[ConfigKey.THROTTLING_CONFIG],
+    ],
+    advanced: [
+      {
+        ...DEFAULT_DATA_OPTIONS,
+        components: [
+          FIELD[ConfigKey.TAGS],
+          FIELD[ConfigKey.APM_SERVICE_NAME],
+          FIELD[ConfigKey.SCREENSHOTS],
+          FIELD[ConfigKey.NAMESPACE],
+        ],
+      },
+    ],
+  },
```
All monitor types have at least two steps. MultiStep is the only monitor type with three steps.

dominiqueclarke replied:
Single is not single in terms of form steps, but in terms of synthetics script steps.

dominiqueclarke commented on x-pack/plugins/synthetics/common/constants/monitor_defaults.ts:
```diff
@@ -158,16 +161,22 @@ export const DEFAULT_FIELDS: MonitorDefaults = {
     ...DEFAULT_HTTP_SIMPLE_FIELDS,
     ...DEFAULT_HTTP_ADVANCED_FIELDS,
     ...DEFAULT_TLS_FIELDS,
+    [ConfigKey.FORM_MONITOR_TYPE]: FormMonitorType.HTTP,
```
This new abstraction is needed because browser monitors are now split into two types, multistep and single. Since this is now part of the monitor type, it needs to be defined on defaults. Likewise, this also ensures that new monitors created from Uptime are created with the appropriate spec to be able to be compatible with Synthetics edit page.

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+      <>
+        <FormattedMessage
+          id="xpack.synthetics.monitorConfig.indexResponseHeaders.helpText"
+          defaultMessage="Controls the indexing of the HTTP response headers to "
```
Is this missing a word or phrase at the end?

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+      <>
+        <FormattedMessage
+          id="xpack.synthetics.monitorConfig.indexResponseBody.helpText"
+          defaultMessage="Controls the indexing of the HTTP response body contents to"
```
Is this missing a word or phrase at the end?

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+    fieldKey: ConfigKey.RESPONSE_HEADERS_CHECK,
+    component: HeaderField,
+    label: i18n.translate('xpack.synthetics.monitorConfig.responseHeadersCheck.label', {
+      defaultMessage: 'Check response headers contain',
```
Is this missing a word or phrase at the end? Or maybe it makes sense in the UI?

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+      defaultMessage: 'Certificate authorities',
+    }),
+    helpText: i18n.translate('xpack.synthetics.monitorConfig.certificateAuthorities.helpText', {
+      defaultMessage: 'PEM formatted custom certificate authorities.',
```
```suggestion
      defaultMessage: 'PEM-formatted custom certificate authorities.',
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+      defaultMessage: 'Client certificate',
+    }),
+    helpText: i18n.translate('xpack.synthetics.monitorConfig.clientCertificate.helpText', {
+      defaultMessage: 'PEM formatted certificate for TLS client authentication.',
```
```suggestion
      defaultMessage: 'PEM-formatted certificate for TLS client authentication.',
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+      defaultMessage: 'Client key',
+    }),
+    helpText: i18n.translate('xpack.synthetics.monitorConfig.clientKey.helpText', {
+      defaultMessage: 'PEM formatted certificate key for TLS client authentication.',
```
```suggestion
      defaultMessage: 'PEM-formatted certificate key for TLS client authentication.',
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/field_config.tsx:
```diff
@@ -0,0 +1,979 @@
+    controlled: true,
+    helpText: i18n.translate('xpack.synthetics.monitorConfig.throttling.helpText', {
+      defaultMessage:
+        'Simulate network throttling (download, upload, latency). More options will be added in a future version.',
```
Is that second sentence necessary? I haven't seen any references to future features elsewhere. 

```suggestion
        'Simulate network throttling (download, upload, latency).',
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/form_config.tsx:
```diff
@@ -0,0 +1,232 @@
+    defaultMessage: 'Data options',
+  }),
+  description: i18n.translate('xpack.synthetics.monitorConfig.section.dataOptions.description', {
+    defaultMessage: 'A description goes here',
```
Is this placeholder text?

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/form/form_config.tsx:
```diff
@@ -0,0 +1,232 @@
+    defaultMessage: 'TLS options',
+  }),
+  description: i18n.translate('xpack.synthetics.monitorConfig.section.tlsOptions.description', {
+    defaultMessage: 'TLS Description',
```
Is this placeholder text?

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/steps/step_config.tsx:
```diff
@@ -0,0 +1,153 @@
+    <EuiFlexItem grow={false}>
+      <EuiButton href={`elastic-synthetics-recorder://`} iconType="popout" iconSide="right">
+        {i18n.translate('xpack.synthetics.monitorConfig.monitorScriptStep.scriptRecorder.launch', {
+          defaultMessage: 'Launch Script Recorder',
```
As far as I know, the official name is "Synthetics Recorder" (see https://github.com/elastic/observability-docs/issues/1907#issuecomment-1169668147). 

```suggestion
          defaultMessage: 'Launch Synthetics Recorder',
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/steps/step_config.tsx:
```diff
@@ -0,0 +1,153 @@
+        {i18n.translate(
+          'xpack.synthetics.monitorConfig.monitorScriptStep.scriptRecorder.download',
+          {
+            defaultMessage: 'Download Script Recorder',
```
```suggestion
            defaultMessage: 'Download Synthetics Recorder',
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/steps/step_config.tsx:
```diff
@@ -0,0 +1,153 @@
+          <p>
+            <FormattedMessage
+              id="xpack.synthetics.monitorConfig.monitorScriptStep.description"
+              defaultMessage="Use Elastic Script Recorder to generate a script and then upload it. Alternatively, you can write your own {playwright} script and paste it in the script editor."
```
```suggestion
              defaultMessage="Use Elastic Synthetics Recorder to generate a script and then upload it. Alternatively, you can write your own {playwright} script and paste it in the script editor."
```

colleenmcginnis replied:
> script editor 

Should this be "script field"?

dominiqueclarke replied:
I think either term works, because we have a full scale code editor inline as a field, editor works as well. What would you prefer?

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/steps/step_config.tsx:
```diff
@@ -0,0 +1,153 @@
+          <p>
+            <FormattedMessage
+              id="xpack.synthetics.monitorConfig.monitorScriptEditStep.description"
+              defaultMessage="Use Elastic Script Recorder to generate and upload a script. Alternatively, you can edit the existing {playwright} script (or paste a new one) in the script editor."
```
```suggestion
              defaultMessage="Use Elastic Synthetics Recorder to generate and upload a script. Alternatively, you can edit the existing {playwright} script (or paste a new one) in the script editor."
```

colleenmcginnis commented on x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_add_edit/fields/uploader.tsx:
```diff
@@ -0,0 +1,96 @@
+);
+
+const PROMPT_TEXT = i18n.translate('xpack.synthetics.monitorConfig.uploader.label', {
+  defaultMessage: 'Select or drag and drop a script.js file',
```
Should we align with the messages below?

```suggestion
  defaultMessage: 'Select or drag and drop a .js file',
```

colleenmcginnis reviewed (commented):
There are so many descriptions in here. 😅  I did my best to find them all in the diff, but let me know if there's anything that needs a closer look. I wasn't able to find all of these descriptions in the context of the UI, so let me know if my comments don't make sense based on the context.

kibanamachine commented:
merge conflict between base and head

bataya0 commented:
If a step needs or uses a parameter,  like a password.
How would the UI generate the parameter logic for the inline script and also how/where would it store the password value securely and get put on the synthetic test location.

We figured this part out in a standalone synthetic setup on a server,  but just wondering how the UI will do the same thing.

awahab07 commented:
When submitting the form to add a monitor, form error message appears, though it doesn't block the form submission. Happens for all monitor types. Doesn't happen on Edit form.

https://user-images.githubusercontent.com/2748376/182391971-7621e501-365d-440a-a174-79bf3157f820.mov

There's a runtime api type error which is not there in `main`. It occurs on the monitor list page.
![Screenshot 2022-08-02 at 15 31 07](https://user-images.githubusercontent.com/2748376/182392919-8e611745-aece-4c60-b857-c46005b072dd.png)

dominiqueclarke commented:
@elasticmachine merge upstream

kibanamachine commented:
merge conflict between base and head

dominiqueclarke commented:
> When submitting the form to add a monitor, form error message appears, though it doesn't block the form submission. Happens for all monitor types. Doesn't happen on Edit form.

The issues with the form have been resolved.

The issue with the runtime type is because we added a new required field `ConfigKey.FORM_MONITOR_TYPE`, to the common fields. I've adjusted this to be optional. However, we may want to consider a saved object migration in the future to make this required. https://github.com/elastic/kibana/pull/132204/commits/f7ad34db28d8514c5b5d5df8eb6177819a399a2a

awahab07 commented:
@elasticmachine merge upstream

dominiqueclarke commented:
@elasticmachine merge upstream

dominiqueclarke commented:
@elasticmachine merge upstream

dominiqueclarke commented:
@elasticmachine merge upstream

dominiqueclarke commented:
@elasticmachine merge upstream

awahab07 commented:
Feature tested the PR again and LGTM.

There's a small enhancement which could be done in a follow up PR. After "Create / Update monitor" button is clicked, it should be disabled (possibly with a spinner) to prevent the user from triggering additional requests.

https://user-images.githubusercontent.com/2748376/185938834-76646dcd-362c-444a-b967-1ebf85fe69ca.mov
